pkg/diff/sql_graph.go

```diff
@@ -65,3 +60,216 @@ func (s *sqlGraph) toOrderedStatements() ([]Statement, error) {
 	}
 	return stmts, nil
 }
+
+// partialSQLGraph represents the SQL statements and their dependencies. Every SQL statement is a
+// vertex in the graph. This is different from a graph because a dependency may exist between a node
+// within this part and a node in another part.
+type partialSQLGraph struct {
+	vertices     []sqlVertex
+	dependencies []dependency
+}
+
+func (s *partialSQLGraph) statements() []Statement {
+	var statements []Statement
+	for _, vertex := range s.vertices {
+		statements = append(statements, vertex.Statements...)
+	}
+	return statements
+}
+
+func concatPartialGraphs(parts ...partialSQLGraph) partialSQLGraph {
+	var vertices []sqlVertex
+	var dependencies []dependency
+	for _, part := range parts {
+		vertices = append(vertices, part.vertices...)
+		dependencies = append(dependencies, part.dependencies...)
+	}
+	return partialSQLGraph{
+		vertices:     vertices,
+		dependencies: dependencies,
+	}
+}
+
+func graphFromPartials(parts partialSQLGraph) (*sqlGraph, error) {
+	graph := newSqlGraph()
+	for _, vertex := range parts.vertices {
+		// It's possible the node already exists. merge it if it does
+		if graph.HasVertexWithId(vertex.GetId()) {
+			vertex = mergeVertices(graph.GetVertex(vertex.GetId()), vertex)
+		}
+		graph.AddVertex(vertex)
+	}
+
+	for _, dep := range parts.dependencies {
+		sourceVertex := sqlVertex{
+			ObjId:      dep.sourceObjId,
+			DiffType:   dep.sourceType,
+			Statements: nil,
+		}
+		targetVertex := sqlVertex{
+			ObjId:      dep.targetObjId,
+			DiffType:   dep.targetType,
+			Statements: nil,
+		}
+
+		// To maintain the correctness of the graph, we will add a dummy vertex for the missing dependencies
+		addVertexIfNotExists(graph, sourceVertex)
+		addVertexIfNotExists(graph, targetVertex)
+
+		if err := graph.AddEdge(sourceVertex.GetId(), targetVertex.GetId()); err != nil {
+			return nil, fmt.Errorf("adding edge from %s to %s: %w", sourceVertex.GetId(), targetVertex.GetId(), err)
+		}
+	}
+
+	return graph, nil
+}
+
+func mergeVertices(old, new sqlVertex) sqlVertex {
+	return sqlVertex{
+		ObjId:      old.ObjId,
+		DiffType:   old.DiffType,
+		Statements: append(old.Statements, new.Statements...),
+	}
+}
+
+func addVertexIfNotExists(graph *sqlGraph, vertex sqlVertex) {
+	if !graph.HasVertexWithId(vertex.GetId()) {
+		graph.AddVertex(vertex)
+	}
+}
+
+type sqlVertexGenerator[S schema.Object, Diff diff[S]] interface {
+	Add(S) (partialSQLGraph, error)
```

Comment:
Now, SQL generators can return entire graphs for any one operation. I.e. when a table is altered, it can return n addresable SQL vertex nodes, one for each column, that SQL vertex generators will be able to globally reference